### PR TITLE
API 2.0 networking refactor

### PR DESCRIPTION
**Pull request by starfi5h**
Major networking refactor by @Reousa 

- Add `ConcurrentPlayerCollection`
- Remove `IPlayerManager` and `PlayerManager`
- Add `IsDedicated`, `IsClient`, `IsServer` properties in `IMultiplayerSession`
- Add `SendToMatching` method in `INetworkProvider`

New mecha combat setting by @sp00ktober in 0.10.x branch

- Add `ConstructionModule`, `FightData`, `UpdateMech` method in `IMechaData`
- Add combat-related upgrade settings in `IPlayerTechBonuses`.

### CODE REVIEW

File: NebulaPatcher/Patches/Dynamic/Player_Patch.cs [hunk 1]
```diff
@@ -24,26 +24,44 @@ public static bool SetSandCount_Prefix(long newSandCount)
...
         switch (Multiplayer.IsActive)
         {
             //Soil should be given in singleplayer or to the host who then syncs it back to all players.
             case true when Multiplayer.Session.LocalPlayer.IsHost:
                 var deltaSandCount = (int)(newSandCount - GameMain.mainPlayer.sandCount);
                 if (deltaSandCount != 0)
                 {
-                    Multiplayer.Session.Network.PlayerManager.UpdateSyncedSandCount(deltaSandCount);
-                    Multiplayer.Session.Network.SendPacket(new PlayerSandCount(newSandCount));
+                    UpdateSyncedSandCount(newSandCount - GameMain.mainPlayer.sandCount);
```
Comment (Zzarek) [inline NebulaPatcher/Patches/Dynamic/Player_Patch.cs:35]: Is this not DeltaSandCount?

File: NebulaPatcher/MonoBehaviours/NebulaBootstrapper.cs [hunk 2]
```diff
@@ -23,7 +24,9 @@ private void LateUpdate()
     {
         if (Multiplayer.IsActive)
         {
-            Multiplayer.Session.Network.Update();
+            //@TODO: Make this a single call
```
Comment (Zzarek) [inline NebulaPatcher/MonoBehaviours/NebulaBootstrapper.cs:27]: Any reason to postpone this ToDo now? Or alternatively remove the ToDo?

File: NebulaAPI/DataStructures/ConcurrentPlayerCollection.cs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using NebulaAPI.GameState;
+using NebulaAPI.Networking;
+
+namespace NebulaAPI.DataStructures;
+
+public class ConcurrentPlayerCollection
+{
```
Comment (Zzarek) [inline NebulaAPI/DataStructures/ConcurrentPlayerCollection.cs:11]: I think the standard C# convention for private classes is to put at the bottom. It doesn't matter that much, but It did feel off when I was reading, though to be fair i don't use them often.

Though the convention does make sense to me. As with every purely style comment anyone ever leaves on a PR, I encourage you to feel absolutely no qualms about closing this comment and ignoring it.

File: NebulaModel/Networking/NebulaConnection.cs [hunk 4]
```diff
@@ -81,24 +103,24 @@ private void OnSendCompleted(bool result)
...
-            {
-                peerSocket.Close((ushort)reason, reasonString);
-            }
-            else
-            {
-                throw new ArgumentException("Reason string cannot take up more than 123 bytes");
-            }
-        }
-    }
+    // public void Disconnect(DisconnectionReason reason = DisconnectionReason.Normal, string reasonString = null)
```
Comment (Zzarek) [inline NebulaModel/Networking/NebulaConnection.cs:106]: Did you forget to delete commented code?

**Review by Zzarek — APPROVED** [review]

**Review by sp00ktober — APPROVED** [review]

**Pull request merged**